plugins/aws_ecr/lib/samson_aws_ecr/samson_plugin.rb

```diff
@@ -3,29 +3,55 @@
 module SamsonAwsEcr
   class Engine < Rails::Engine
     AMAZON_REGISTRY = /\A.*\.dkr.ecr.(?<region>[\w\-]+).amazonaws.com\z/
-    mattr_accessor :ecr_client
-    mattr_accessor(:credentials_expire_at) { Time.current }
 
-    initializer :ecr_client do
-      if match = AMAZON_REGISTRY.match(ENV['DOCKER_REGISTRY'])
-        SamsonAwsEcr::Engine.ecr_client = Aws::ECR::Client.new(region: match['region'])
+    class << self
+      # we make sure the repo exists so pushes do not fail
+      # - ignores if the repo already exists
+      # - ignores if the repo cannot be created due to permission problems
+      def ensure_repository(repository)
+        return unless ecr_client
+        name = repository.split('/', 2).last
+        ecr_client.describe_repositories(repository_names: [name])
+      rescue Aws::ECR::Errors::RepositoryNotFoundException
+        ecr_client.create_repository(repository_name: name)
```

Comment:
On Tue, Jun 14, 2016 at 03:58:59PM -0700, Michael Grosser wrote:
> >  
> > -    initializer :ecr_client do
> > -      if match = AMAZON_REGISTRY.match(ENV['DOCKER_REGISTRY'])
> > -        SamsonAwsEcr::Engine.ecr_client = Aws::ECR::Client.new(region: match['region'])
> > +    class << self
> > +      # we make sure the repo exists so pushes do not fail
> > +      # - ignores if the repo already exists
> > +      # - ignores if the repo cannot be created due to permission problems
> > +      def ensure_repository(repository)
> > +        return unless ecr_client
> > +        name = repository.split('/', 2).last
> > +        ecr_client.describe_repositories(repository_names: [name])
> > +      rescue Aws::ECR::Errors::RepositoryNotFoundException
> > +        ecr_client.create_repository(repository_name: name)
> 
> it is possible, but in both cases it will end up in a AccessDenied error
> which is caught below

Ohh, I wasn't sure if in Ruby that will be caught. Great! (I will need to read
something about ruby, I can't avoid it for much longer :-D)

> added docs for needed permissions

Awesome! :)